### PR TITLE
[AMORO-2299]: Cancel the running optimizing process from ams web console

### LINKED ISSUE #2299: [Subtask]: Cancel the running optimizing process from ams web console
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### Description

In the current amoro version, we can not kill the running optimizing process in ams web console when the external optimizer is canceled by us.

### Parent issue

- https://github.com/NetEase/amoro/issues/1862

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by tcodehuber**
## Why are the changes needed?

- close #2299 
This is a subtask of #1862.
In this PR, the AMS will use a 'Cancel' button to support killing the running process, as belows:
screenshot-20231113-185457

## Brief change log

- define a new function in interface TableService :  TableRuntime getRuntime(TableIdentifier tableIdentifier)
- add a button in 'Optimizing' Tab

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/table/TableService.java [hunk 1]
```diff
@@ -84,6 +84,8 @@ public interface TableService extends CatalogService, TableManager {
    */
   List<TableIDWithFormat> listTables(String catalogName, String dbName);
 
+  TableRuntime getRuntime(TableIdentifier tableIdentifier);
+
```
Comment (XBaith) [inline ams/server/src/main/java/com/netease/arctic/server/table/TableService.java:88]: I'm not sure if the interface matches the design of the `TableService`.

Comment (wangtaohz) [reply ams/server/src/main/java/com/netease/arctic/server/table/TableService.java:88]: We already have a interface `com.netease.arctic.server.table.TableManager#getRuntime`, there is no need to add a new one.

**Comment by baiyangtx** [comment]
Your PR includes frontend changes. It would be helpful to add some screenshots in the comments for reviewers to evaluate the visual effects.

**Comment by wangtaohz** [comment]
> Your PR includes frontend changes. It would be helpful to add some screenshots in the comments for reviewers to evaluate the visual effects.

I find the change of Web UI from the description of this PR.
image

However, this `Cancel` button doesn't seem to inform the user what will be canceled when I click it. Perhaps we should add this button in each line of process, so that users can clearly know which process will be canceled when click it.

Another issue is that to prevent accidental touches, a confirmation window is required when clicking this button.

File: ams/server/src/main/java/com/netease/arctic/server/table/TableService.java (hunk 1 above)
Comment (tcodehuber) [reply ams/server/src/main/java/com/netease/arctic/server/table/TableService.java:88]: I have reviewed com.netease.arctic.server.table.TableManager#getRuntime, which confused me is that we can not build ServerTableIdentifier only with catalog/db/tab from frontend request params.  
We can not get TableRuntime from tableRuntimeMap (private final Map<ServerTableIdentifier, TableRuntime> tableRuntimeMap ), because ServerTableIdentifier's equals method also contains tableId except catalog/db/tab.

**Comment by tcodehuber** [comment]
> > Your PR includes frontend changes. It would be helpful to add some screenshots in the comments for reviewers to evaluate the visual effects.
> 
> I find the change of Web UI from the description of this PR. image
> 
> However, this `Cancel` button doesn't seem to inform the user what will be canceled when I click it. Perhaps we should add this button in each line of process, so that users can clearly know which process will be canceled when click it.
> 
> Another issue is that to prevent accidental touches, a confirmation window is required when clicking this button.



> > Your PR includes frontend changes. It would be helpful to add some screenshots in the comments for reviewers to evaluate the visual effects.
> 
> I find the change of Web UI from the description of this PR. image
> 
> However, this `Cancel` button doesn't seem to inform the user what will be canceled when I click it. Perhaps we should add this button in each line of process, so that users can clearly know which process will be canceled when click it.
> 
> Another issue is that to prevent accidental touches, a confirmation window is required when clicking this button.

That is good, i will add the confirm operation later. In addition, I want to know whether only one running optimizing process can in the RUNNING state in a certain table's the optimizing tab. If that is true, put on a global operation button may be enough.

File: ams/server/src/main/java/com/netease/arctic/server/table/TableService.java (hunk 1 above)
Comment (wangtaohz) [reply ams/server/src/main/java/com/netease/arctic/server/table/TableService.java:88]: Yes, `ServerTableIdentifier` contains a field `tableId` to distinguish tables with the same catalog/db/table, perhaps it is a table been dropped and then recreated.

I think we can add an interface(method) to `com.netease.arctic.server.table.TableService` that can get `ServerTableIdentifier` from `TableIdentifier`, and this method has already been implemented

**Comment by wangtaohz** [comment]
I suggest to put the cancel button in the optimizing detail page,and this page will be introduced after this issue https://github.com/NetEase/amoro/issues/2115.
image

**Comment by baiyangtx** [comment]
> I suggest to put the cancel button in the optimizing detail page,and this page will be introduced after this issue #2115. image

Personally, I don't recommend waiting for another uncertain PR for an already completed PR. 

In terms of this interaction logic, I personally think it is acceptable, and I suggest merging this PR once it is completed. If there is a better solution for the interaction after the other PR is merged, a new PR can be raised to address it.

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java [hunk 2]
```diff
@@ -552,6 +553,30 @@ public void getTableBranches(Context ctx) {
...
+  /**
+   * cancel the running optimizing process of one certain table.
+   *
+   * @param ctx - context for handling the request and response
+   */
+  public void cancelOptimizingProcess(Context ctx) {
+    String catalog = ctx.pathParam("catalog");
+    String db = ctx.pathParam("db");
+    String table = ctx.pathParam("table");
+    Preconditions.checkArgument(
```
Comment (baiyangtx) [inline ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java:565]: From a parameter design perspective, it is best to specify the process ID. This avoids mistakenly canceling an unintended process due to delays between the frontend page and backend data.

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/DashboardServer.java [hunk 3]
```diff
@@ -213,6 +213,9 @@ public EndpointGroup endpoints() {
             get(
                 "/tables/catalogs/{catalog}/dbs/{db}/tables/{table}/branches",
                 tableController::getTableBranches);
+            post(
+                "/tables/catalogs/{catalog}/dbs/{db}/tables/{table}/cancel-process",
+                tableController::cancelOptimizingProcess);
```
[the remark this answers is not in the document]
Comment (tcodehuber) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/DashboardServer.java:218]: > This method also needs to be added to the open-api.

Added, details can be found in new committed code

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java (hunk 2 above)
Comment (tcodehuber) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java]: > From a parameter design perspective, it is best to specify the process ID. This avoids mistakenly canceling an unintended process due to delays between the frontend page and backend data.

Added, details can be found in new committed code

**Comment by tcodehuber** [comment]
> > Your PR includes frontend changes. It would be helpful to add some screenshots in the comments for reviewers to evaluate the visual effects.
> 
> I find the change of Web UI from the description of this PR. image
> 
> However, this `Cancel` button doesn't seem to inform the user what will be canceled when I click it. Perhaps we should add this button in each line of process, so that users can clearly know which process will be canceled when click it.
> 
> Another issue is that to prevent accidental touches, a confirmation window is required when clicking this button.

I have added a modal confirm action, later we can move this action into optimizing process details page.

**Comment by tcodehuber** [comment]
> > I suggest to put the cancel button in the optimizing detail page,and this page will be introduced after this issue #2115. image
> 
> Personally, I don't recommend waiting for another uncertain PR for an already completed PR.
> 
> In terms of this interaction logic, I personally think it is acceptable, and I suggest merging this PR once it is completed. If there is a better solution for the interaction after the other PR is merged, a new PR can be raised to address it.

Yes, I have tested it. This PR now is ready to merge.

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java [hunk 4]
```diff
@@ -606,6 +602,36 @@ public void getTableBranches(Context ctx) {
...
+        StringUtils.isNotBlank(catalog)
+            && StringUtils.isNotBlank(db)
+            && StringUtils.isNotBlank(table),
+        "catalog.database.tableName can not be empty in any element");
+    Preconditions.checkState(tableService.catalogExist(catalog), "invalid catalog!");
+
+    ServerTableIdentifier serverTableIdentifier =
+        tableService.getServerTableIdentifier(
+            TableIdentifier.of(catalog, db, table).buildTableIdentifier());
+    TableRuntime tableRuntime = tableService.getRuntime(serverTableIdentifier);
```
Comment (baiyangtx) [inline ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java:625]: serverTableIdentifier maybe null if the table doesn't have a TableRuntime.

File: ams/server/src/main/java/com/netease/arctic/server/table/TableService.java [hunk 5]
```diff
@@ -85,6 +85,8 @@ public interface TableService extends CatalogService, TableManager {
    */
   List<TableIDWithFormat> listTables(String catalogName, String dbName);
 
+  ServerTableIdentifier getServerTableIdentifier(TableIdentifier id);
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/table/TableService.java:88]: Add a comment for this method.

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java (hunk 4 above)
Comment (tcodehuber) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/TableController.java]: > serverTableIdentifier maybe null if the table doesn't have a TableRuntime.

Done, plz review it again, thx.

File: ams/server/src/main/java/com/netease/arctic/server/table/TableService.java (hunk 5 above)
Comment (tcodehuber) [reply ams/server/src/main/java/com/netease/arctic/server/table/TableService.java:93]: > Add a comment for this method.

done.

**Review by wangtaohz — APPROVED** [review]
LGTM

**Review by XBaith — APPROVED** [review]
LGTM👏

**Review by zhoujinsong — COMMENTED** [review]
@fantasyni I left a suggestion about the new API, please take a look.

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/DashboardServer.java [hunk 6]
```diff
@@ -214,6 +214,9 @@ public EndpointGroup endpoints() {
             get(
                 "/tables/catalogs/{catalog}/dbs/{db}/tables/{table}/branches",
                 tableController::getTableBranches);
+            post(
```
[the remark this answers is not in the document]
Comment (wangtaohz) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/DashboardServer.java:217]: I think `cancel` is okay, and the reason why using`tables/catalogs/{catalog}/dbs/{db}/tables/{table}/optimizing-processes/{processId}/cancel` instead of `/tables/catalogs/{catalog}/dbs/{db}/tables/{table}/cancel-process/{processId}` is to be a RESTful API.

[the remark this answers is not in the document]
Comment (tcodehuber) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/DashboardServer.java:217]: OK, i will use "/tables/catalogs/{catalog}/dbs/{db}/tables/{table}/cancel-process/{processId}" instead.

[the remark this answers is not in the document]
Comment (tcodehuber) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/DashboardServer.java:217]: I have changed the api, plz review it again.

**Review by zhoujinsong — APPROVED** [review]
LGTM.
